### PR TITLE
Remove ear clipping algorithm

### PR DESCRIPTION
**Pull request by TomFischer**
- New implementation of `createSurface` that uses Gmsh for surface generation
- Ear clipping algorithm is not necessary anymore

### CODE REVIEW

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        return;
+    }
+
+    // create new GEOObjects and insert a copy of the polyline
+    auto polyline_points = std::make_unique<std::vector<GeoLib::Point*>>();
+    GeoLib::GEOObjects geo;
+    auto ply_points = ply.getPointsVec();
+    for (auto p : ply_points)
+        polyline_points->push_back(new GeoLib::Point(*p));
+    std::string ply_name = "test_ply";
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.cpp]: Why is "test_ply" hard coded?

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    geo.addPointVec(std::move(polyline_points), ply_name);
+    auto polyline = new GeoLib::Polyline(*geo.getPointVec(ply_name));
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.cpp]: Minor: Could be unique_ptr, later `release()`.

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    char file_base_name_c[L_tmpnam];
+    if (! std::tmpnam(file_base_name_c))
+    {
+       OGS_FATAL("Could not create unique file name.");
+    }
+    std::string const file_base_name(file_base_name_c);
+    gmsh_io.writeToFile(file_base_name + ".geo");
+    std::string gmsh_command =
+        "\"gmsh\" -2 -algo meshadapt " + file_base_name + ".geo";
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.cpp]: Why the quotes aroung gmsh? Does that work on windows?
This command might be problematic: file_base_name is not escaped. If it contains spaces, it might be interpreted as several cmdline args.

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    gmsh_command += " -o " + file_base_name + ".msh";
+    system(gmsh_command.c_str());
+    auto surface_mesh =
+        FileIO::GMSH::readGMSHMesh(file_base_name + ".msh");
+
+    std::remove(std::string(file_base_name + ".geo").c_str());
+    std::remove(std::string(file_base_name + ".msh").c_str());
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.cpp]: Minor: `std::string` probably not necessary here.

File: Applications/FileIO/Legacy/createSurface.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Copyright (c) 2012-2017, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#pragma once
+
+#include <vector>
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.h]: Not needed I guess.

File: Applications/FileIO/Legacy/createSurface.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+
+namespace GeoLib
+{
+class Polyline;
+class Surface;
```
Comment (chleh) [inline Applications/FileIO/Legacy/createSurface.h]: not needed.

File: Applications/FileIO/Legacy/OGSIOVer4.cpp [hunk 7]
```diff
@@ -402,22 +404,14 @@ std::string readSurfaces(std::istream &in,
...
-                          pnt_vec,
-                          path,
-                          errors);
+        tag = readSurface(in, polygon_vec, sfc_vec, sfc_names, ply_vec,
+                          ply_vec_names, pnt_vec, path, errors);
         if (n_polygons < polygon_vec.size())
         {
-            // subdivide polygon in simple polygons
-            GeoLib::Surface* sfc(GeoLib::Surface::createSurface(
+            std::string surface_name("test");
```
Comment (chleh) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: Why is "test" hardcoded?

File: Applications/FileIO/Legacy/OGSIOVer4.cpp [hunk 8]
```diff
@@ -402,22 +404,14 @@ std::string readSurfaces(std::istream &in,
...
+            FileIO::createSurface(
                 *(dynamic_cast<GeoLib::Polyline*>(
-                    polygon_vec[polygon_vec.size() - 1]))));
-            sfc_vec.push_back(sfc);
+                    polygon_vec[polygon_vec.size() - 1])), surface_name, geo);
```
Comment (chleh) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: Minor: `polygon_vec.back()`

File: Applications/FileIO/Legacy/OGSIOVer4.cpp [hunk 9]
```diff
@@ -475,32 +475,39 @@ bool readGLIFileV4(const std::string& fname,
     else
         INFO("GeoLib::readGLIFile(): tag #POLYLINE not found.");
 
+    if (!ply_vec->empty())
+        geo.addPolylineVec(
+            std::move(ply_vec), unique_name,
+            std::move(ply_names));  // KR: insert into GEOObjects if not empty
```
Comment (chleh) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: The comment just says what is already clear from the code. But what does the comment mean?

File: Applications/FileIO/Legacy/OGSIOVer4.cpp [hunk 10]
```diff
@@ -475,32 +475,39 @@ bool readGLIFileV4(const std::string& fname,
...
     {
         INFO("GeoLib::readGLIFile(): read surfaces from stream.");
+
         tag = readSurfaces(in,
                            *sfc_vec,
                            *sfc_names,
-                           *ply_vec,
-                           *ply_names,
+                           *geo.getPolylineVec(unique_name),
+                           ply_names_copy,
```
Comment (chleh) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: Why is the copy necessary now?
Is there a problem now, if ply_vec is empty after reading?

File: Applications/DataExplorer/DataView/GEOModels.h [hunk 11]
```diff
@@ -76,6 +76,9 @@ public slots:
     void appendSurfaceVec(std::string const& name);
     void removeSurfaceVec(std::string const& name);
 
+    void renameGeometry(std::string const& old_name,
```
Comment (chleh) [inline Applications/DataExplorer/DataView/GEOModels.h]: No Documentation, but maybe thats not mandatory here...

File: Tests/MeshLib/TestBoundaryElementSearch.cpp [hunk 12]
```diff
@@ -131,43 +131,40 @@ TEST_F(MeshLibBoundaryElementSearchInSimpleHexMesh, SurfaceSearch)
...
+    MeshGeoToolsLib::BoundaryElementsSearcher boundary_element_searcher(
+        *_hex_mesh, mesh_node_searcher);
+    std::vector<MeshLib::Element*> const& found_faces_sfc_b(
+        boundary_element_searcher.getBoundaryElements(sfc_bottom));
     ASSERT_EQ(n_eles_2d, found_faces_sfc_b.size());
-    double sum_area_b = std::accumulate(found_faces_sfc_b.begin(), found_faces_sfc_b.end(), 0.0,
-                [](double v, MeshLib::Element*e){return v+e->getContent();});
+    double sum_area_b = std::accumulate(
+        found_faces_sfc_b.begin(), found_faces_sfc_b.end(), 0.0,
+        [](double v, MeshLib::Element* e) { return v + e->getContent(); });
```
Comment (chleh) [inline Tests/MeshLib/TestBoundaryElementSearch.cpp]: Btw: Since now we are dealing with nonlinear geometry regularly, I assume that `e->getContent()` needs an update. This is not related to this PR. @endJunction, please keep it in mind.

File: Applications/FileIO/Legacy/OGSIOVer4.cpp (hunk 10 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: Since `ply_names` is a `auto_ptr` and is given to the `GEOObject` object `geo` it is not usable anymore.  For this reason a copy is necessary.

File: Applications/FileIO/Legacy/createSurface.cpp (hunk 1 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/createSurface.cpp]: Changed to `temporary_polyline_name` to reflect better that this is a temporary name used only in this function.

File: Applications/FileIO/Legacy/createSurface.cpp (hunk 3 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/createSurface.cpp]: Added `\"` around filenames.

File: Applications/FileIO/Legacy/createSurface.h (hunk 5 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/createSurface.h]: Removed.

File: Applications/FileIO/Legacy/createSurface.h (hunk 6 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/createSurface.h]: Removed.

**Comment by chleh** [comment]
Shell escaping:
https://stackoverflow.com/a/288071
https://stackoverflow.com/a/3669819
https://stackoverflow.com/a/8539271
TL;DR: There exist some rather easy solutions for escaping. Using `fork()` and `exec()` might be better, but they are available only on *nix. So decision about that issue is ...?

File: GeoLib/GEOObjects.h [hunk 13]
```diff
@@ -241,6 +242,9 @@ class GEOObjects final
      */
     int mergeGeometries(std::vector<std::string> const & names, std::string &merged_geo_name);
 
+    void renameGeometry(std::string const& old_name,
```
Comment (endJunction) [inline GeoLib/GEOObjects.h]: Interface method should be documented.

File: Applications/DataExplorer/DataView/GEOModels.h (hunk 11 above)
Comment (endJunction) [inline Applications/DataExplorer/DataView/GEOModels.h]: I'd rather like to see docu of the GeoObjects::renameGeometry interface (mentioned below)...

File: GeoLib/GEOObjects.h (hunk 13 above)
Comment (TomFischer) [inline GeoLib/GEOObjects.h]: I thought the name is self documenting.

File: Applications/FileIO/Legacy/OGSIOVer4.cpp [hunk 14]
```diff
@@ -487,31 +475,42 @@ bool readGLIFileV4(const std::string& fname,
     else
         INFO("GeoLib::readGLIFile(): tag #POLYLINE not found.");
 
+    if (!ply_vec->empty())
+        geo.addPolylineVec(std::move(ply_vec), unique_name,
+                           std::move(ply_names));
+
+    // copy of ply_names
+    std::map<std::string, std::size_t> ply_names_copy;
```
Comment (endJunction) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: minor: Better would be a reason for a copy. Otherwise remove the comment---it's already in the variable name.

File: Applications/FileIO/Legacy/OGSIOVer4.cpp (hunk 10 above)
Comment (endJunction) [inline Applications/FileIO/Legacy/OGSIOVer4.cpp]: Ah, I see some explanation here; Put it in the comment above, please.

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    if (! std::tmpnam(file_base_name_c))
+    {
+       OGS_FATAL("Could not create unique file name.");
+    }
+    std::string const file_base_name(file_base_name_c);
+    gmsh_io.writeToFile(file_base_name + ".geo");
+    std::string gmsh_command =
+        "gmsh -2 -algo meshadapt \"" + file_base_name + ".geo\"";
+    gmsh_command += " -o \"" + file_base_name + ".msh\"";
+    system(gmsh_command.c_str());
```
Comment (endJunction) [inline Applications/FileIO/Legacy/createSurface.cpp]: std::system

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    auto surface_mesh =
+        FileIO::GMSH::readGMSHMesh("\"" + file_base_name + ".msh\"");
+
+    std::remove((file_base_name + ".geo").c_str());
+    std::remove((file_base_name + ".msh").c_str());
```
Comment (endJunction) [inline Applications/FileIO/Legacy/createSurface.cpp]: Return values should be checked

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+       OGS_FATAL("Could not create unique file name.");
+    }
+    std::string const file_base_name(file_base_name_c);
+    gmsh_io.writeToFile(file_base_name + ".geo");
+    std::string gmsh_command =
+        "gmsh -2 -algo meshadapt \"" + file_base_name + ".geo\"";
+    gmsh_command += " -o \"" + file_base_name + ".msh\"";
+    system(gmsh_command.c_str());
+    auto surface_mesh =
+        FileIO::GMSH::readGMSHMesh("\"" + file_base_name + ".msh\"");
```
Comment (endJunction) [inline Applications/FileIO/Legacy/createSurface.cpp]: surface_mesh could be a nullptr...

File: Applications/FileIO/Legacy/createSurface.cpp [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+    std::remove((file_base_name + ".geo").c_str());
+    std::remove((file_base_name + ".msh").c_str());
+
+    // convert the surface mesh into a geometric surface
+    MeshLib::convertMeshToGeo(*surface_mesh, geometries,
+                              std::numeric_limits<double>::epsilon());
```
Comment (endJunction) [inline Applications/FileIO/Legacy/createSurface.cpp]: conversion also could go wrong; convertMeshToGeo returns a boolean...

File: Applications/FileIO/Legacy/createSurface.h [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace GeoLib
+{
+class Polyline;
+class GEOObjects;
+}
+
+namespace FileIO
+{
+void createSurface(GeoLib::Polyline const& ply,
```
Comment (endJunction) [inline Applications/FileIO/Legacy/createSurface.h]: Short docu needed.

File: Applications/FileIO/SHPInterface.cpp [hunk 20]
```diff
@@ -176,17 +178,10 @@ void SHPInterface::readPolygons(const SHPHandle &hSHP, int numberOfElements, con
...
-    for (auto const* polyline : *polylines) {
-        GeoLib::Surface* sfc(GeoLib::Surface::createSurface(*polyline));
-        if (sfc)
-            sfc_vec->push_back(sfc);
-        else {
-            WARN("SHPInterface::readPolygons(): Could not triangulate polygon.")
-        }
+    for (auto const* polyline : *polylines)
+    {
+        FileIO::createSurface(*polyline, _geoObjects, listName);
```
Comment (endJunction) [inline Applications/FileIO/SHPInterface.cpp]: the createSurface could fail in multiple ways. I'd keep the testing; maybe return a flag from createSurface...

File: Applications/Utils/GeoTools/TriangulatePolyline.cpp [hunk 21]
```diff
@@ -97,17 +98,8 @@ int main(int argc, char *argv[])
...
-
-    GeoLib::SurfaceVec* sfc_vec (geo_objects.getSurfaceVecObj(geo_names[0]));
-    if (sfc_vec == nullptr)
-    {
-        geo_objects.addSurfaceVec(std::move(new_sfc), geo_names[0]);
-        sfc_vec = geo_objects.getSurfaceVecObj(geo_names[0]);
-    }
-    else
-        geo_objects.appendSurfaceVec(*new_sfc, geo_names[0]);
+    FileIO::createSurface(*line, geo_objects, geo_names[0]);
```
Comment (endJunction) [inline Applications/Utils/GeoTools/TriangulatePolyline.cpp]: Same here, the createSurface could fail...

**Comment by chleh** [comment]
:+1: But in the future we should have a bulletproof shell escaping algorithm in OGS.

File: GeoLib/GEOObjects.h (hunk 13 above)
Comment (endJunction) [inline GeoLib/GEOObjects.h]: Does the renaming always work? Are there any preconditions?

File: Applications/FileIO/Legacy/createSurface.cpp (hunk 15 above)
Comment (TomFischer) [inline Applications/FileIO/Legacy/createSurface.cpp]: Checking the return value is not obvious. The cppreference documentation states that `std::system` returns an implementation-defined value.

**Comment by TomFischer** [comment]
Rework is finished. Thanks for the helpful review. I hope I addressed all points.

File: Applications/DataExplorer/DataView/GEOModels.h (hunk 11 above)
Comment (chleh) [reply Applications/DataExplorer/DataView/GEOModels.h]: @endJunction: Almost none of the methods in this class are documented. So personally I'd rather not be too picky here.

Comment (endJunction) [reply Applications/DataExplorer/DataView/GEOModels.h]: Yes, not here. That's fine. ;)
Tom added the docu at the right place, where most functions are documented. :+1: 
:stop_button:

**Comment by endJunction** [comment]
Merged it too early. Had to reset the master branch. Sorry for inconvenience.
